### PR TITLE
Validate cluster member names

### PR DESCRIPTION
**Pull request by MggMuggins**
Added a system test for the example which covers this PR; happy to rebase that out of this PR and iterate on it separately if that's preferred.

Should we make any consideration for existing clusters here? This shouldn't cause any issues for them based on where I'm doing the validation, but is it worth spitting out a warning at startup for an invalid member name?

### CODE REVIEW

**Review by roosterfish — COMMENTED** [review]
Looks good to me. 

Besides blocking new clusters/members from using invalid hostnames, should we also add a change allowing existing members with invalid names to be removed from a cluster as reported in #57 ?

**Comment by masnax** [comment]
> Looks good to me.
> 
> Besides blocking new clusters/members from using invalid hostnames, should we also add a change allowing existing members with invalid names to be removed from a cluster as reported in #57 ?

Yeah, I think that's a good point. We should fix the node removal process to allow nodes with invalid names to be removed.

Additionally, I think we should wait to merge this until #86 is merged so we can add an API extension and prevent a daemon with hostname validation from clustering with an older daemon.

**Comment by MggMuggins** [comment]
`gorilla/mux` [interprets `%2F` as `/` in routes](https://gist.github.com/MggMuggins/43eaca212402a5c4bb4aed34d6a0afa4) (doesn't seem to matter if [this](https://github.com/canonical/microcluster/blob/main/internal/daemon/daemon.go#L299) is set or not), so any request for a node name containing a slash returns a 404 to the client and never gets to its handler function (e.g. [`clusterMemberDelete`](https://github.com/canonical/microcluster/blob/main/internal/rest/resources/cluster.go#L319) is never called). [Bug](https://github.com/gorilla/mux/issues/639).

Here's three bad ideas:
- Base64 the node name and add a query param to indicate the encoding, as in `/cluster/bWljb2Nsb3VkLzE0Cg==?name=base64`
  - This would allow a greater set of node names, although I don't think that is particularly desirable
- Use the database ID instead of a name, and again use a query param to indicate it: `/cluster/2?name=id`
- Use `cluster/{name:.+}` [here](https://github.com/canonical/microcluster/blob/main/internal/rest/resources/cluster.go#L44), which would prevent adding any endpoints under `cluster/{name}/...`

Anyone have any thoughts?

**Comment by tomponline** [comment]
> `gorilla/mux` [interprets `%2F` as `/` in routes](https://gist.github.com/MggMuggins/43eaca212402a5c4bb4aed34d6a0afa4) (doesn't seem to matter if [this](https://github.com/canonical/microcluster/blob/main/internal/daemon/daemon.go#L299) is set or not), so any request for a node name containing a slash returns a 404 to the client and never gets to its handler function (e.g. [`clusterMemberDelete`](https://github.com/canonical/microcluster/blob/main/internal/rest/resources/cluster.go#L319) is never called). [Bug](https://github.com/gorilla/mux/issues/639).
> 
> Here's three bad ideas:
> 
>     * Base64 the node name and add a query param to indicate the encoding, as in `/cluster/bWljb2Nsb3VkLzE0Cg==?name=base64`
>       
>       * This would allow a greater set of node names, although I don't think that is particularly desirable
> 
>     * Use the database ID instead of a name, and again use a query param to indicate it: `/cluster/2?name=id`
> 
>     * Use `cluster/{name:.+}` [here](https://github.com/canonical/microcluster/blob/main/internal/rest/resources/cluster.go#L44), which would prevent adding any endpoints under `cluster/{name}/...`
> 
> 
> Anyone have any thoughts?

LXD uses `mux.UseEncodedPath() // Allow encoded values in path segments.` As we do expect to get `/` for snapshot names in path segments.

**Comment by tomponline** [comment]
https://github.com/canonical/lxd/pull/10276

**Comment by MggMuggins** [comment]
`mux` was working just fine. I did a poor job of debugging on the client side; it was sending the [wrong URL](https://github.com/canonical/microcluster/pull/106/commits/d4f9600a2d54d966aaa08151a4ddaf16fd7bc5fe).

I've added the validation logic on the server side as well; atm it's very unlikely that anyone is using the API without `microcluster/microcluster` but the server should be doing the validation. I left the client side validation in because it will give a better error message than `Error: Failed to join cluster with the given join token`. Happy for feedback here.

Per @roosterfish and @masnax' comments, is it even possible that a member with a slash in the name has fully joined the cluster? Every cluster member has a trust store in its state dir with a yaml file for each other member; members with a slash in their name would cause the path to be something like `truststore/c/4.yaml`; since `truststore/c` doesn't exist, the leader is unable to fully join the new member, so the new member stays `PENDING` in the db. The error in #57 is what happens when a pending member is removed (see [here](https://github.com/canonical/microcluster/blob/main/internal/rest/resources/cluster.go#L340)). I don't see any hooks running before the failure in my reproducer, so is there anything else that's needed other than removing the db entry?

**Comment by masnax** [comment]
> Per @roosterfish and @masnax' comments, is it even possible that a member with a slash in the name has fully joined the cluster? Every cluster member has a trust store in its state dir with a yaml file for each other member; members with a slash in their name would cause the path to be something like `truststore/c/4.yaml`; since `truststore/c` doesn't exist, the leader is unable to fully join the new member, so the new member stays `PENDING` in the db. 

Hmm, yes this means that these clusters are all fundamentally broken anyway, there was never a possibility of inter-cluster communication.

I'm not sure anymore if it would make sense to have a special case to remove these perpetually pending cluster members from the database, it's impossible that any real-world setup got very far without becoming non-functional.

**Comment by roosterfish** [comment]
> Per @roosterfish and @masnax' comments, is it even possible that a member with a slash in the name has fully joined the cluster?

That is a valid point I haven't thought about when writing the comment, fine for me to leave it as is.

**Comment by tomponline** [comment]
needs a rebase

**Comment by MggMuggins** [comment]
Blocked on #86 per @masnax; I was planning to wait and do all those changes at the same time

**Review by masnax — CHANGES_REQUESTED** [review]
Since we decided against making changes to the cluster removal process (and since any cluster that's set up with an invalid name is fundamentally broken), I think we can go ahead and merge this once the changes are addressed, without waiting for #86.

File: internal/rest/resources/control.go
Comment (MggMuggins) [reply internal/rest/resources/control.go]: Duh, thanks.

File: microcluster/app.go
Comment (MggMuggins) [reply microcluster/app.go]: Wasn't concerned about efficiency, just the error message. The above fixes that.

**Review by roosterfish — APPROVED** [review]

**Review by masnax — APPROVED** [review]

**Pull request merged**
